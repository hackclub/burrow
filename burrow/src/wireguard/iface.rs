use std::{net::IpAddr, ops::Deref, sync::Arc};

use anyhow::Error;
use fehler::throws;
use futures::future::join_all;
use ip_network_table::IpNetworkTable;
use tokio::sync::{Notify, RwLock};
use tracing::{debug, error};
use tun::tokio::TunInterface;

use super::{noise::Tunnel, Peer, PeerPcb};

pub struct IndexedPcbs {
    pcbs: Vec<Arc<PeerPcb>>,
    allowed_ips: IpNetworkTable<usize>,
}

impl Default for IndexedPcbs {
    fn default() -> Self {
        Self::new()
    }
}

impl IndexedPcbs {
    pub fn new() -> Self {
        Self {
            pcbs: vec![],
            allowed_ips: IpNetworkTable::new(),
        }
    }

    pub fn insert(&mut self, pcb: PeerPcb) {
        let idx: usize = self.pcbs.len();
        for allowed_ip in pcb.allowed_ips.iter() {
            self.allowed_ips.insert(*allowed_ip, idx);
        }
        self.pcbs.insert(idx, Arc::new(pcb));
    }

    pub fn find(&self, addr: IpAddr) -> Option<usize> {
        let (_, &idx) = self.allowed_ips.longest_match(addr)?;
        Some(idx)
    }
}

impl FromIterator<PeerPcb> for IndexedPcbs {
    fn from_iter<I: IntoIterator<Item = PeerPcb>>(iter: I) -> Self {
        iter.into_iter().fold(Self::new(), |mut acc, pcb| {
            acc.insert(pcb);
            acc
        })
    }
}

enum IfaceStatus {
    Running,
    Idle,
}

pub struct Interface {
    pub tun: Arc<RwLock<Option<TunInterface>>>,
    pub pcbs: Arc<IndexedPcbs>,
    status: Arc<RwLock<IfaceStatus>>,
    stop_notifier: Arc<Notify>,
}

async fn is_running(status: Arc<RwLock<IfaceStatus>>) -> bool {
    let st = status.read().await;
    matches!(st.deref(), IfaceStatus::Running)
}

impl Interface {
    #[throws]
    pub fn new<I: IntoIterator<Item = Peer>>(peers: I) -> Self {
        let pcbs: IndexedPcbs = peers
            .into_iter()
            .map(PeerPcb::new)
            .collect::<Result<_, _>>()?;

        let pcbs = Arc::new(pcbs);
        Self {
            pcbs,
            tun: Arc::new(RwLock::new(None)),
            status: Arc::new(RwLock::new(IfaceStatus::Idle)),
            stop_notifier: Arc::new(Notify::new()),
        }
    }

    pub async fn set_tun(&self, tun: TunInterface) {
        debug!("Setting tun interface");
        self.tun.write().await.replace(tun);
        let mut st = self.status.write().await;
        *st = IfaceStatus::Running;
    }

    pub fn get_tun(&self) -> Arc<RwLock<Option<TunInterface>>> {
        self.tun.clone()
    }

    pub async fn remove_tun(&self) {
        let mut st = self.status.write().await;
        self.stop_notifier.notify_waiters();
        *st = IfaceStatus::Idle;
    }

    pub async fn run(&self) -> anyhow::Result<()> {
        let pcbs = self.pcbs.clone();
        let tun = self.tun.clone();
        let status = self.status.clone();
        let stop_notifier = self.stop_notifier.clone();
        log::info!("Starting interface");

        let outgoing = async move {
            while is_running(status.clone()).await {
                let mut buf = [0u8; 3000];

                let src = {
                    let t = tun.read().await;
                    let Some(_tun) = t.as_ref() else {
                        continue;
                    };
                    tokio::select! {
                        _ = stop_notifier.notified() => continue,
                        pkg = _tun.recv(&mut buf[..]) => match pkg {
                            Ok(len) => &buf[..len],
                            Err(e) => {
                                error!("Failed to read from interface: {}", e);
                                continue
                            }
                        },
                    }
                };

                let dst_addr = match Tunnel::dst_address(src) {
                    Some(addr) => addr,
                    None => {
                        debug!("No destination found");
                        continue
                    }
                };

                debug!("Routing packet to {}", dst_addr);

                let Some(idx) = pcbs.find(dst_addr) else {
                    continue
                };

                debug!("Found peer:{}", idx);

                match pcbs.pcbs[idx].send(src).await {
                    Ok(..) => {
                        let addr = pcbs.pcbs[idx].endpoint;
                        debug!("Sent packet to peer {}", addr);
                    }
                    Err(e) => {
                        log::error!("Failed to send packet {}", e);
                        continue
                    }
                };
            }
        };

        let mut tsks = vec![];
        let tun = self.tun.clone();
        let outgoing = tokio::task::spawn(outgoing);
        tsks.push(outgoing);
        debug!("preparing to spawn read tasks");

        {
            let pcbs = &self.pcbs;
            for i in 0..pcbs.pcbs.len() {
                debug!("spawning read task for peer {}", i);
                let pcb = pcbs.pcbs[i].clone();
                let tun = tun.clone();
                let main_tsk = async move {
                    if let Err(e) = pcb.open_if_closed().await {
                        log::error!("failed to open pcb: {}", e);
                        return
                    }
                    let r2 = pcb.run(tun).await;
                    if let Err(e) = r2 {
                        log::error!("failed to run pcb: {}", e);
                    } else {
                        debug!("pcb ran successfully");
                    }
                };

                let pcb = pcbs.pcbs[i].clone();
                let status = self.status.clone();
                let update_timers_tsk = async move {
                    let mut buf = [0u8; 65535];
                    while is_running(status.clone()).await {
                        tokio::time::sleep(tokio::time::Duration::from_millis(250)).await;
                        match pcb.update_timers(&mut buf).await {
                            Ok(..) => (),
                            Err(e) => {
                                error!("Failed to update timers: {}", e);
                                return
                            }
                        }
                    }
                };

                let pcb = pcbs.pcbs[i].clone();
                let status = self.status.clone();
                let reset_rate_limiter_tsk = async move {
                    while is_running(status.clone()).await {
                        tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
                        pcb.reset_rate_limiter().await;
                    }
                };
                tsks.extend(vec![
                    tokio::spawn(main_tsk),
                    tokio::spawn(update_timers_tsk),
                    tokio::spawn(reset_rate_limiter_tsk),
                ]);
                debug!("task made..");
            }
            debug!("spawned read tasks");
        }
        debug!("preparing to join..");
        join_all(tsks).await;
        debug!("joined!");
        Ok(())
    }
}
