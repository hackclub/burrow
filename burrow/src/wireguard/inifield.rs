use std::str::FromStr;

use anyhow::{Error, Result};

pub struct IniField(String);

impl FromStr for IniField {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self(s.to_string()))
    }
}

impl TryFrom<IniField> for Vec<String> {
    type Error = Error;

    fn try_from(field: IniField) -> Result<Self, Self::Error> {
        Ok(field.0.split(',').map(|s| s.trim().to_string()).collect())
    }
}

impl TryFrom<IniField> for u32 {
    type Error = Error;

    fn try_from(value: IniField) -> Result<Self, Self::Error> {
        value.0.parse().map_err(Error::from)
    }
}

impl TryFrom<IniField> for Option<u32> {
    type Error = Error;

    fn try_from(value: IniField) -> Result<Self, Self::Error> {
        if value.0.is_empty() {
            Ok(None)
        } else {
            value.0.parse().map(Some).map_err(Error::from)
        }
    }
}

impl TryFrom<IniField> for String {
    type Error = Error;

    fn try_from(value: IniField) -> Result<Self, Self::Error> {
        Ok(value.0)
    }
}

impl TryFrom<IniField> for Option<String> {
    type Error = Error;

    fn try_from(value: IniField) -> Result<Self, Self::Error> {
        if value.0.is_empty() {
            Ok(None)
        } else {
            Ok(Some(value.0))
        }
    }
}

impl<T> TryFrom<Option<T>> for IniField
where
    T: ToString,
{
    type Error = Error;

    fn try_from(value: Option<T>) -> Result<Self, Self::Error> {
        Ok(match value {
            Some(v) => Self(v.to_string()),
            None => Self(String::new()),
        })
    }
}

impl IniField {
    fn new(value: &str) -> Self {
        Self(value.to_string())
    }
}
