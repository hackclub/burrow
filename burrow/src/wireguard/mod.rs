pub mod config;
mod iface;
mod inifield;
mod noise;
mod pcb;
mod peer;

pub use config::Config;
pub use iface::Interface;
pub use pcb::PeerPcb;
pub use peer::Peer;
