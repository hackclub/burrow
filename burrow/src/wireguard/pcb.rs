use std::{net::SocketAddr, sync::Arc};
use std::time::Duration;

use anyhow::{Error, Result};
use fehler::throws;
use ip_network::IpNetwork;
use rand::random;
use tokio::{net::UdpSocket, sync::RwLock, task::JoinHandle, time::timeout};
use tokio::io::AsyncWrite;
use tun::tokio::TunInterface;
use crate::wireguard::noise::errors::WireGuardError;

use super::{
    noise::{TunnResult, Tunnel},
    Peer,
};

#[derive(Debug)]
pub struct PeerPcb {
    pub endpoint: SocketAddr,
    pub allowed_ips: Vec<IpNetwork>,
    pub handle: RwLock<Option<JoinHandle<()>>>,
    socket: RwLock<Option<UdpSocket>>,
    tunnel: RwLock<Tunnel>,
}

impl PeerPcb {
    #[throws]
    pub fn new(peer: Peer) -> Self {
        let tunnel = RwLock::new(
            Tunnel::new(
                peer.private_key,
                peer.public_key,
                peer.preshared_key,
                None,
                1,
                None,
            )
            .map_err(|s| anyhow::anyhow!("{}", s))?,
        );
        Self {
            endpoint: peer.endpoint,
            allowed_ips: peer.allowed_ips,
            handle: RwLock::new(None),
            socket: RwLock::new(None),
            tunnel,
        }
    }

    pub async fn open_if_closed(&self) -> Result<(), Error> {
        if self.socket.read().await.is_none() {
            let socket = UdpSocket::bind("0.0.0.0:0").await?;
            socket.connect(self.endpoint).await?;
            self.socket.write().await.replace(socket);
        }
        Ok(())
    }

    pub async fn run(&self, tun_interface: Arc<RwLock<TunInterface>>) -> Result<(), Error> {
        tracing::debug!("starting read loop for pcb... for {:?}", &self);
        let rid: i32 = random();
        let mut buf: [u8; 3000] = [0u8; 3000];
        tracing::debug!("start read loop {}", rid);
        loop {
            tracing::debug!("{}: waiting for packet", rid);
            let guard = self.socket.read().await;
            let Some(socket) = guard.as_ref() else {
                self.open_if_closed().await?;
                continue
            };
            let mut res_buf = [0; 1500];
            // tracing::debug!("{} : waiting for readability on {:?}", rid, socket);
            let len = match socket.recv(&mut res_buf).await {
                Ok(l) => l,
                Err(e) => {
                    log::error!("{}: error reading from socket: {:?}", rid, e);
                    continue
                }
            };
            let mut res_dat = &res_buf[..len];
            tracing::debug!("{}: Decapsulating {} bytes", rid, len);
            tracing::debug!("{:?}", &res_dat);
            loop {
                match self
                    .tunnel
                    .write()
                    .await
                    .decapsulate(None, res_dat, &mut buf[..])
                {
                    TunnResult::Done => break,
                    TunnResult::Err(e) => {
                        tracing::error!(message = "Decapsulate error", error = ?e);
                        break
                    }
                    TunnResult::WriteToNetwork(packet) => {
                        tracing::debug!("WriteToNetwork: {:?}", packet);
                        self.open_if_closed().await?;
                        self.socket.read().await.as_ref().unwrap().send(packet).await?;
                        tracing::debug!("WriteToNetwork done");
                        res_dat = &[];
                        continue
                    }
                    TunnResult::WriteToTunnelV4(packet, addr) => {
                        tracing::debug!("WriteToTunnelV4: {:?}, {:?}", packet, addr);
                        tun_interface.read().await.send(packet).await?;
                        break
                    }
                    TunnResult::WriteToTunnelV6(packet, addr) => {
                        tracing::debug!("WriteToTunnelV6: {:?}, {:?}", packet, addr);
                        tun_interface.read().await.send(packet).await?;
                        break
                    }
                }
            }
        }
    }

    pub async fn send(&self, src: &[u8]) -> Result<(), Error> {
        tracing::debug!("Sending packet: {:?}", src);
        let mut dst_buf = [0u8; 3000];
        match self.tunnel.write().await.encapsulate(src, &mut dst_buf[..]) {
            TunnResult::Done => {
                tracing::debug!("Encapsulate done");
            }
            TunnResult::Err(e) => {
                tracing::error!(message = "Encapsulate error", error = ?e)
            }
            TunnResult::WriteToNetwork(packet) => {
                self.open_if_closed().await?;
                let handle = self.socket.read().await;
                let Some(socket) = handle.as_ref() else {
                    tracing::error!("No socket for peer");
                    return Ok(())
                };
                tracing::debug!("Our Encapsulated packet: {:?}", packet);
                socket.send(packet).await?;
            }
            _ => panic!("Unexpected result from encapsulate"),
        };
        Ok(())
    }

    pub async fn update_timers(&self, dst: &mut [u8]) -> Result<(), Error> {
        match self.tunnel.write().await.update_timers(dst) {
            TunnResult::Done => {}
            TunnResult::Err(WireGuardError::ConnectionExpired) => {
            }
            TunnResult::Err(e) => {
                tracing::error!(message = "Update timers error", error = ?e)
            }
            TunnResult::WriteToNetwork(packet) => {
                tracing::debug!("Sending Packet for timer update: {:?}", packet);
                self.open_if_closed().await?;
                let handle = self.socket.read().await;
                let Some(socket) = handle.as_ref() else {
                    tracing::error!("No socket for peer");
                    return Ok(())
                };
                socket.send(packet).await?;
                tracing::debug!("Sent Packet for timer update");
            }
            _ => panic!("Unexpected result from update_timers"),
        };
        Ok(())
    }

    pub async fn reset_rate_limiter(&self) {
        self.tunnel.read().await.reset_rate_limiter();
    }
}
