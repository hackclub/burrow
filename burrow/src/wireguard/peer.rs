use std::{fmt, net::SocketAddr};

use anyhow::Error;
use fehler::throws;
use ip_network::IpNetwork;
use x25519_dalek::{PublicKey, StaticSecret};

pub struct Peer {
    pub endpoint: SocketAddr,
    pub private_key: StaticSecret,
    pub public_key: PublicKey,
    pub allowed_ips: Vec<IpNetwork>,
}

impl fmt::Debug for Peer {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Peer")
            .field("endpoint", &self.endpoint)
            .field("public_key", &self.public_key)
            .field("allowed_ips", &self.allowed_ips)
            .finish()
    }
}
