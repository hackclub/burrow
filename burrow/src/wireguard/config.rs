use std::{net::ToSocketAddrs, str::FromStr};

use anyhow::{anyhow, Error, Result};
use base64::{engine::general_purpose, Engine};
use fehler::throws;
use ini::{Ini, Properties};
use ip_network::IpNetwork;
use serde::{Deserialize, Serialize};
use x25519_dalek::{PublicKey, StaticSecret};

use super::inifield::IniField;
use crate::wireguard::{Interface as WgInterface, Peer as WgPeer};

#[throws]
fn parse_key(string: &str) -> [u8; 32] {
    let value = general_purpose::STANDARD.decode(string)?;
    let mut key = [0u8; 32];
    key.copy_from_slice(&value[..]);
    key
}

#[throws]
fn parse_secret_key(string: &str) -> StaticSecret {
    let key = parse_key(string)?;
    StaticSecret::from(key)
}

#[throws]
fn parse_public_key(string: &str) -> PublicKey {
    let key = parse_key(string)?;
    PublicKey::from(key)
}

/// A raw version of Peer Config that can be used later to reflect configuration files.
/// This should be later converted to a `WgPeer`.
/// Refers to https://github.com/pirate/wireguard-docs?tab=readme-ov-file#overview
#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct Peer {
    pub public_key: String,
    pub preshared_key: Option<String>,
    pub allowed_ips: Vec<String>,
    pub endpoint: String,
    pub persistent_keepalive: Option<u32>,
    pub name: Option<String>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct Interface {
    pub private_key: String,
    pub address: Vec<String>,
    pub listen_port: Option<u32>,
    pub dns: Vec<String>,
    pub mtu: Option<u32>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct Config {
    #[serde(rename = "Peer")]
    pub peers: Vec<Peer>,
    pub interface: Interface, // Support for multiple interfaces?
}

impl TryFrom<Config> for WgInterface {
    type Error = anyhow::Error;

    fn try_from(cfig: Config) -> Result<Self, Error> {
        let sk = parse_secret_key(&cfig.interface.private_key)?;
        let wg_peers: Vec<WgPeer> = cfig
            .peers
            .iter()
            .map(|p| {
                Ok(WgPeer {
                    private_key: sk.clone(),
                    public_key: parse_public_key(&p.public_key)?,
                    endpoint: p
                        .endpoint
                        .to_socket_addrs()?
                        .find(|sock| sock.is_ipv4())
                        .ok_or(anyhow!("DNS Lookup Fails!"))?,
                    preshared_key: match &p.preshared_key {
                        None => Ok(None),
                        Some(k) => parse_key(k).map(Some),
                    }?,
                    allowed_ips: p
                        .allowed_ips
                        .iter()
                        .map(|ip_addr| {
                            IpNetwork::from_str(ip_addr)
                                .map_err(|e| anyhow!("Error parsing IP Network {}: {}", ip_addr, e))
                        })
                        .collect::<Result<Vec<IpNetwork>>>()?,
                })
            })
            .collect::<Result<Vec<WgPeer>>>()?;
        WgInterface::new(wg_peers)
    }
}

impl Default for Config {
    fn default() -> Self {
        Self {
            interface: Interface {
                private_key: "OEPVdomeLTxTIBvv3TYsJRge0Hp9NMiY0sIrhT8OWG8=".into(),
                address: vec!["10.13.13.2/24".into()],
                listen_port: Some(51820),
                dns: Default::default(),
                mtu: Default::default(),
            },
            peers: vec![Peer {
                endpoint: "wg.burrow.rs:51820".into(),
                allowed_ips: vec!["8.8.8.8/32".into(), "0.0.0.0/0".into()],
                public_key: "8GaFjVO6c4luCHG4ONO+1bFG8tO+Zz5/Gy+Geht1USM=".into(),
                preshared_key: Some("ha7j4BjD49sIzyF9SNlbueK0AMHghlj6+u0G3bzC698=".into()),
                persistent_keepalive: Default::default(),
                name: Default::default(),
            }],
        }
    }
}

fn props_get<T>(props: &Properties, key: &str) -> Result<T>
where
    T: TryFrom<IniField, Error = anyhow::Error>,
{
    IniField::try_from(props.get(key))?.try_into()
}

impl TryFrom<&Properties> for Interface {
    type Error = anyhow::Error;

    fn try_from(props: &Properties) -> Result<Self, Error> {
        Ok(Self {
            private_key: props_get(props, "PrivateKey")?,
            address: props_get(props, "Address")?,
            listen_port: props_get(props, "ListenPort")?,
            dns: props_get(props, "DNS")?,
            mtu: props_get(props, "MTU")?,
        })
    }
}

impl TryFrom<&Properties> for Peer {
    type Error = anyhow::Error;

    fn try_from(props: &Properties) -> Result<Self, Error> {
        Ok(Self {
            public_key: props_get(props, "PublicKey")?,
            preshared_key: props_get(props, "PresharedKey")?,
            allowed_ips: props_get(props, "AllowedIPs")?,
            endpoint: props_get(props, "Endpoint")?,
            persistent_keepalive: props_get(props, "PersistentKeepalive")?,
            name: props_get(props, "Name")?,
        })
    }
}

impl Config {
    pub fn from_toml(toml: &str) -> Result<Self> {
        toml::from_str(toml).map_err(Into::into)
    }

    pub fn from_ini(ini: &str) -> Result<Self> {
        let ini = Ini::load_from_str(ini)?;
        let interface = ini
            .section(Some("Interface"))
            .ok_or(anyhow!("Interface section not found"))?;
        let peers = ini.section_all(Some("Peer"));
        Ok(Self {
            interface: Interface::try_from(interface)?,
            peers: peers
                .into_iter()
                .map(|v| Peer::try_from(v))
                .collect::<Result<Vec<Peer>>>()?,
        })
    }

    pub fn from_content_fmt(content: &str, fmt: &str) -> Result<Self> {
        match fmt {
            "toml" => Self::from_toml(content),
            "ini" | "conf" => Self::from_ini(content),
            _ => Err(anyhow::anyhow!("Unsupported format: {}", fmt)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tst_config_toml() {
        let cfig = Config::default();
        let toml = toml::to_string(&cfig).unwrap();
        println!("{}", &toml);
        insta::assert_snapshot!(toml);
        let cfig2: Config = toml::from_str(&toml).unwrap();
        assert_eq!(cfig, cfig2);
    }
}
