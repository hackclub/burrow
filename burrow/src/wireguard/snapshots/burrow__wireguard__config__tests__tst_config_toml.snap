---
source: burrow/src/wireguard/config.rs
expression: toml
---
[[Peer]]
public_key = "8GaFjVO6c4luCHG4ONO+1bFG8tO+Zz5/Gy+Geht1USM="
preshared_key = "ha7j4BjD49sIzyF9SNlbueK0AMHghlj6+u0G3bzC698="
allowed_ips = ["8.8.8.8/32", "0.0.0.0/0"]
endpoint = "wg.burrow.rs:51820"

[interface]
private_key = "OEPVdomeLTxTIBvv3TYsJRge0Hp9NMiY0sIrhT8OWG8="
address = ["10.13.13.2/24"]
listen_port = 51820
dns = []

